modules/storages/app/contracts/storages/storages/base_contract.rb

```diff
@@ -0,0 +1,107 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2022 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require 'net/http'
+require 'uri'
+
+# Purpose: common functionalities shared by CreateContract and UpdateContract
+# UpdateService by default checks if UpdateContract exists
+# and uses the contract to validate the model under consideration
+# (normally it's a model).
+module Storages::Storages
+  class BaseContract < ::ModelContract
+    MINIMAL_NEXTCLOUD_VERSION = 23
+
+    include ::Storages::Storages::Concerns::ManageStoragesGuarded
+    include ActiveModel::Validations
+
+    attribute :name
+    validates :name, length: { minimum: 1, maximum: 255 }, allow_nil: false
+
+    attribute :provider_type
+    validates :provider_type, inclusion: { in: ->(*) { Storages::Storage::PROVIDER_TYPES } }, allow_nil: false
+
+    attribute :creator, writable: false do
+      validate_creator_is_user
+    end
+
+    attribute :host
+    validates :host, length: { minimum: 1, maximum: 255 }, allow_nil: false
+    validates_url :host
```

Comment:
We only have access to this since we require `openid_connect` which in turn requires `validate_url`. If we use it within our code, we should add that dependency to our `Gemfile`

Reply:
Since we already validate for it to be a url, I think we don't need the other checks above, based on the the specs within the gem: https://github.com/perfectline/validates_url/blob/master/spec/validate_url_spec.rb

However, what `validates_url` will allow are query props like in `https://something.com/blubs?abc=true` which will lead to failures within the reachability checks.

Reply:
> We only have access to this since we require openid_connect which in turn requires validate_url. If we use it within our code, we should add that dependency to our Gemfile

It is also used in [`Webhooks::Webhook`](https://github.com/opf/openproject/blob/cade7fdaabd47e2805c2b80dd67c107744293f96/modules/webhooks/app/models/webhooks/webhook.rb#L9) through [another syntax](https://github.com/perfectline/validates_url/tree/master).

> However, what `validates_url` will allow are query props like in `https://something.com/blubs?abc=true` which will lead to failures within the reachability checks.

Indeed. Is it something we want to guard against?